changelog.d/12738.misc

```diff
@@ -0,0 +1 @@
+Change error message when attempting to login to a non-existent account to match that returned when the account exists but the password is invalid. Contributed by Daniel Aloni.
```

Comment:
I think the intent is the following (please confirm @Danieloni1)

```suggestion
Report login failures due to unknown third party identifiers in the same way as failures due to invalid passwords. This prevents an attacker from using the error response to determine if the identifier exists. Contributed by Daniel Aloni.
```

If this really is the intention of this change, I have a few requests:

1. In both error cases, the message ought to be e.g. "Invalid username or password". Consider a user who correctly types their password but misspells their email address. This change would mislead them with the error message `Invalid password`. I believe this is fairly common practice, but I don't have a citation for you. See perhaps https://security.stackexchange.com/questions/62937/should-i-let-a-user-know-they-have-entered-an-unknown-username-or-email-address or https://security.stackexchange.com/questions/62661/generic-error-message-for-wrong-password-or-username-is-this-really-helpful for more insight.
2. To ensure that we really do return the same error code in both cases, these two `LoginError(...)` expressions
    https://github.com/matrix-org/synapse/blob/677c7637283f9da991273953eb2f561d367b595b/synapse/handlers/auth.py#L1207
    https://github.com/matrix-org/synapse/blob/677c7637283f9da991273953eb2f561d367b595b/synapse/handlers/auth.py#L1333
    would ideally be factored out to a module-level constant, e.g. `INVALID_USERNAME_OR_PASSWORD`. I suppose this might cause pain if some exception handler later chose to modify this exception though...

Reply:
@DMRobertson I agree with both suggestions. Just to make sure:

1. Do you want to change every occurence of `Invalid password` across the project? Or just the ones from `synapse/handlers/auth.py`?
2. Should the constant (`INVALID_USERNAME_OR_PASSWORD`) reside in `synapse/api/errors.py` in a new class (say `Messages`), or remain in the file?

Reply:
1. I think just `synapse/handlers.auth.py`. The other uses of `"Invalid password"`I can  see are tests, creating an account, changing a password, or the admin API; there aren't any other public-facing login endpoints. 
2. Just this file please: we're trying to ensure this endpoint is consistent in its responses; I wouldn't expect other parts of the source tree to worry about this detail.

Reply:
Done ✅

Reply:
Thanks. For completeness, I just noticed that Dirk made the same suggestion here, https://github.com/matrix-org/synapse/pull/12738#issuecomment-1128587798.